p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,282 @@
+package libp2pwebrtc
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"os"
+
+	// "io"
+	"net"
+
+	"sync"
+	"time"
+
+	"sync/atomic"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/pion/webrtc/v3"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+)
+
+var _ network.MuxedStream = &dataChannel{}
+
+const defaultReadBufferLen = 4096
+
+// Package pion detached data channel into a net.Conn
+// and then a network.MuxedStream
+type dataChannel struct {
+	// TODO: Are these circular references okay?
+	channel       *webrtc.DataChannel
+	laddr         net.Addr
+	raddr         net.Addr
+	readDeadline  *deadline
+	writeDeadline *deadline
+
+	closeWriteOnce sync.Once
+	closeReadOnce  sync.Once
+	resetOnce      sync.Once
+
+	remoteWriteClosed uint32
+	localWriteClosed  uint32
+
+	remoteReadClosed uint32
+	localReadClosed  uint32
+
+	ctx        context.Context
+	cancel     context.CancelFunc
+	m          sync.Mutex
+	readBuf    bytes.Buffer
+	readSignal chan struct{}
+}
+
+func newDataChannel(
+	channel *webrtc.DataChannel,
+	pc *webrtc.PeerConnection,
+	laddr, raddr net.Addr) *dataChannel {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	result := &dataChannel{
+		channel:       channel,
+		laddr:         laddr,
+		raddr:         raddr,
+		readDeadline:  newDeadline(),
+		writeDeadline: newDeadline(),
+		ctx:           ctx,
+		cancel:        cancel,
+		readSignal:    make(chan struct{}),
+	}
+	result.readBuf.Grow(defaultReadBufferLen)
+
+	channel.OnMessage(result.handleMessage)
+
+	return result
+}
+
+func (d *dataChannel) handleMessage(msg webrtc.DataChannelMessage) {
+	if msg.IsString {
```

Comment:
DataChannels support string type messages an binary messages. For our use case, we only expect binary messages. This would likely be a misbehaving peer.

Reply:
How do they differ on the wire? Can you provide any link so I can read about that?

Reply:
https://www.rfc-editor.org/rfc/rfc9260.html#name-payload-protocol-identifier They are identical to binary messages but have a different payload protocol identifier. https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/media/sctp/dcsctp_transport.cc;l=71;drc=976e95ee408c73c80f9c3c9a0d800025e8374f43

Reply:
The quoted section of the RFC doesn't say anything about string messages.

If we're dealing with a misbehaving peer, we should close the connection, not just ignore it.